command/agent/cache/lease_cache.go

```diff
@@ -170,40 +192,60 @@ func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 		return nil, err
 	}
 
-	// Grab a read lock for this particular request
-	idLock := locksutil.LockForKey(c.idLocks, id)
+	// Check the inflight cache to see if there are other inflight requests
+	// of the same kind, based on the computed ID. If so, we increment a counter
 
-	idLock.RLock()
-	unlockFunc := idLock.RUnlock
-	defer func() { unlockFunc() }()
+	var inflight *inflightRequest
 
-	// Check if the response for this request is already in the cache
-	sendResp, err := c.checkCacheForRequest(id)
-	if err != nil {
-		return nil, err
-	}
-	if sendResp != nil {
-		c.logger.Debug("returning cached response", "path", req.Request.URL.Path)
-		return sendResp, nil
-	}
+	defer func() {
+		// Cleanup on the cache if there are no remaining inflight requests.
+		// This is the last step, so we defer the call first
+		if inflight != nil && inflight.remaining.Load() == 0 {
```

Comment:
When would `inflight` be `nil` at this point?

Reply:
It _shouldn't_ be nil because the conditional further down always assigns this values to something, but I'm nil-checking just in case since this within a defer that's called right after the variable is declared but before the value is assigned.